Added support for number type attributes

dw2 opened the pull request:
This change will allow model attributes with type 'number' to work with the select view. As an added bonus, this will also allow for mixed option sets (i.e. `[[0,'First'], 2, 3, [4,'Fourth'], 5, ...]`).

dw2 commented:
This resolves issue #66

cdaringe commented:
thanks @dw2!  i now see what you're sayin. :)

wraithgar and I had a long thread with another user about permitting the use of mixed types.  at current time, we are hesistant to support multiple option types in the options set, per discussion [here](https://github.com/AmpersandJS/ampersand-select-view/issues/64#issuecomment-176277662).  also, mixing of some types are just plain incompatible.  however, we definitely do want to support numbers only.  can't believe this has never come up before!

very grateful for your PR.  for the time being, would you be kind enough to:

1. add a test?  they should be relatively straightforward. to imitate.  see the test folder
2. find the matching option within common types?  if you want to use `.find` and be more functional about it, thats ok too. :)

dw2 commented:
1. Sure, I can add a test. I'll try to use the other tests as an example.
2. Can you explain what you mean by "common types" and using ".find"? Do you mean adding checks for other attribute types, like booleans and dates? Something else?

cdaringe commented:
> Can you explain what you mean by "common types"

apologies for the lack of clarity.  "find the matching option within common types?" should have read "adjust the code to find the matching option assuming that the options in `this.options` are homogeneous.  e.g. 'remove support for mixed option sets'."

>  and using ".find"?

if you don't want to use the status quo `for` loop, you're welcome to find the matching option using `this.options.find(function(option) { ... })`.  just conveying no need to maintain the status quo if you didnt care to.

dw2 commented:
I removed support for mixed option types and added a test for numbers when using `[[number','label'], ...` option types. However I am unable to run the test locally because I cannot get zuul installed properly to run the test suite.

wraithgar commented:
checked out branch as of 82ad163 and the tests pass locally for me so let's not let that be a blocker here, I'm currently trying to help doug get zuul running via other messaging channels.

dw2 commented:
@wraithgar helped me out via Twitter DM. I can run tests locally now, but am still having separate issues with zuul. For now, Taking this as a win. Thank you @wraithgar and @cdaringe for the help and feedback. Let me know if you need anything else for this PR. I hope to contribute more in the future.

dw2 commented:
Also, thank you @jweakley for helping me troubleshoot NPM issues.

wraithgar commented on ampersand-select-view.js:
```diff
@@ -359,15 +359,26 @@ module.exports = View.extend({
+                    }
                 }
+            } else {
+                throw new Error('no options provided');
```
I like and support this, we tend to want to throw exceptions when devs get themselves into places that would give unexpected results.  This will mean a major release though because it would not have thrown in this case before.

dw2 replied:
Do you think I should remove this for now and do a separate PR for it?

wraithgar replied:
Nope! Leave it. Versions are cheap I'm just making sure we see that so we don't miss it.

wraithgar commented on ampersand-select-view.js:
```diff
@@ -359,15 +359,26 @@ module.exports = View.extend({
+            if (i >= 0) {
+                // find value in collection as [['val','text'], ...]
+                if (Array.isArray(this.options[0])) {
+                    for (i; i >= 0; i--) {
```
Reusing the variable `i` here seems like the wrong move, it confuses its purpose imo.

Honestly we could simplify it by scoping it (`var i;`) and then the if statement `i >= 0` can simply be `this.options.length >= 1`

Typically the `length` attribute is cast into its own var as an optimization for loops. It's only being referenced once during the if statement so it's really not necessary.

dw2 replied:
I did this to avoid the `'i' is already defined.` error that the pre-commit hook throws. What do you think about this as a solution? Declaring i at the top of the function, and then setting it for each loop:

```
    getOptionByValue: function(value) {
        var model, i;
        if (this.options.isCollection) {
            // find value in collection, error if no model found
            if (this.options.indexOf(value) === -1) model = this.getModelForId(value);
            else model = value;
            if (!model) throw new Error('model or model idAttribute not found in options collection');
            return this.yieldModel ? model : model[this.idAttribute];
        } else if (Array.isArray(this.options)) {
            if (this.options.length > 0) {
                // find value in collection as [['val','text'], ...]
                if (Array.isArray(this.options[0])) {
                    for (i = this.options.length - 1; i >= 0; i--) {
                        if (this.options[i][0] == value) {
                            return this.options[i][0];
                        }
                    }
                // find value in collection as ['val', ...]
                } else {
                    for (i = this.options.length - 1; i >= 0; i--) {
                        if (this.options[i] == value) {
                            return this.options[i];
                        }
                    }
                }
            } else {
                throw new Error('no options provided');
            }
            throw new Error('value not in set of provided options');
        }
        throw new Error('select option set invalid');
    },
```

dw2 replied:
Another option would be to use a different var name for each loop.

wraithgar replied:
Your code example is exactly what I was trying to suggest, that will work, or using two variable names will work, not picky either way honestly.
